### ISSUE TITLE
Problems launching Gearpump in YARN environment

### ISSUE THREAD

**Issue by kkasravi** [gearpump/gearpump#1817]
Output is 

```
$ output/target/pack/bin/yarnclient launch -package /user/gearpump/gearpump-2.11-0.7.3-SNAPSHOT.zip -output /tmp/cluster.conf
[INFO] [12/31/2015 14:20:19.231] [ClusterConfig$] loading config file application.conf...
[WARN] [12/31/2015 14:20:20.464] [NativeCodeLoader] Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[WARN] [12/31/2015 14:20:20.488] [DomainSocketFactory] The short-circuit local reads feature cannot be used because libhadoop cannot be loaded.
[INFO] [12/31/2015 14:20:20.505] [RMProxy] Connecting to ResourceManager at cdh-master-0.node.demogotapaas.consul/10.10.10.146:8032
[INFO] [12/31/2015 14:20:20.632] [YarnClient] Starting YarnClient...
[INFO] [12/31/2015 14:20:20.912] [Slf4jLogger] Slf4jLogger started
[INFO] [12/31/2015 14:20:21.029] [Remoting] Starting remoting
[INFO] [12/31/2015 14:20:21.216] [Remoting] Remoting started; listening on addresses :[akka.tcp://launchCluster@127.0.0.1:58363]
[INFO] [12/31/2015 14:20:21.280] [Metrics$] Metrics is enabled...,  true
[INFO] [12/31/2015 14:20:21.293] [LaunchCluster] Starting AM
[INFO] [12/31/2015 14:20:21.633] [YarnClient] Create application, appId: application_1450294801051_10164
[INFO] [12/31/2015 14:20:21.634] [LaunchCluster] Uploading configuration files to remote HDFS(under hdfs://nameservice1/user/ec2-user/.gearpump_app10164/conf/)...
[INFO] [12/31/2015 14:20:22.098] [YarnClient] Submit Application application_1450294801051_10164 to YARN...
[INFO] [12/31/2015 14:20:22.131] [YarnClientImpl] Submitted application application_1450294801051_10164
[INFO] [12/31/2015 14:20:22.131] [LaunchCluster] Waiting application to finish...
.......
[INFO] [12/31/2015 14:20:29.147] [LaunchCluster] Application application_1450294801051_10164 finished with state RUNNING at 0, info: 
================================================
==Application Id: application_1450294801051_10164
[INFO] [12/31/2015 14:20:29.147] [LaunchCluster] Trying to download active configuration to output path: /tmp/cluster.conf
[INFO] [12/31/2015 14:20:29.147] [LaunchCluster] Resolving YarnAppMaster ActorRef for application application_1450294801051_10164
[INFO] [12/31/2015 14:20:32.195] [HttpMethodDirector] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[INFO] [12/31/2015 14:20:32.196] [HttpMethodDirector] Retrying request
[INFO] [12/31/2015 14:20:32.196] [HttpMethodDirector] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[INFO] [12/31/2015 14:20:32.196] [HttpMethodDirector] Retrying request
[INFO] [12/31/2015 14:20:32.197] [HttpMethodDirector] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[INFO] [12/31/2015 14:20:32.197] [HttpMethodDirector] Retrying request
Failed to connect YarnAppMaster(tried 1)... Connection refused
[INFO] [12/31/2015 14:20:35.200] [HttpMethodDirector] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[INFO] [12/31/2015 14:20:35.200] [HttpMethodDirector] Retrying request
[INFO] [12/31/2015 14:20:35.200] [HttpMethodDirector] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[INFO] [12/31/2015 14:20:35.201] [HttpMethodDirector] Retrying request
[INFO] [12/31/2015 14:20:35.201] [HttpMethodDirector] I/O exception (java.net.ConnectException) caught when processing request: Connection refused
[INFO] [12/31/2015 14:20:35.201] [HttpMethodDirector] Retrying request
Failed to connect YarnAppMaster(tried 2)... Connection refused
[INFO] [12/31/2015 14:20:38.787] [AppMasterResolver] Successfully resolved AppMaster address: akka.tcp://GearpumpAM@127.0.0.1:36624/user/$a
[WARN] [12/31/2015 14:20:38.877] [ReliableDeliverySupervisor] Association with remote system [akka.tcp://GearpumpAM@127.0.0.1:36624] has failed, address is now gated for [5000] ms. Reason: [Association failed with [akka.tcp://GearpumpAM@127.0.0.1:36624]] Caused by: [Connection refused: /127.0.0.1:36624]

Help: 
-package (required:false, default:)<Please specify the gearpump.zip package path on HDFS. If not specified, we will use default value /user/gearpump/gearpump.zip>
-name (required:false, default:Gearpump)<Application name showed in YARN>
-verbose (required:false, default:false)<print verbose log on console>
-output (required:false, default:)<output path for configuration file>

Exception in thread "main" akka.pattern.AskTimeoutException: Ask timed out on [Actor[akka.tcp://GearpumpAM@127.0.0.1:36624/user/$a]] after [15000 ms]
	at akka.pattern.PromiseActorRef$$anonfun$1.apply$mcV$sp(AskSupport.scala:334)
	at akka.actor.Scheduler$$anon$7.run(Scheduler.scala:117)
	at scala.concurrent.Future$InternalCallbackExecutor$.unbatchedExecute(Future.scala:599)
	at scala.concurrent.BatchingExecutor$class.execute(BatchingExecutor.scala:109)
	at scala.concurrent.Future$InternalCallbackExecutor$.execute(Future.scala:597)
	at akka.actor.LightArrayRevolverScheduler$TaskHolder.executeTask(Scheduler.scala:467)
	at akka.actor.LightArrayRevolverScheduler$$anon$8.executeBucket$1(Scheduler.scala:419)
	at akka.actor.LightArrayRevolverScheduler$$anon$8.nextTick(Scheduler.scala:423)
	at akka.actor.LightArrayRevolverScheduler$$anon$8.run(Scheduler.scala:375)
	at java.lang.Thread.run(Thread.java:745)
```

Only one worker is launched. Additional notes:

experiments/yarn/README.md - step 3: 
```
Put the YARN configurations under classpath. Before calling "yarnclient launch", make sure you have put all yarn configuration files under classpath. Typically, you can just copy all files under $HADOOP_HOME/etc/hadoop from one of the YARN Cluster machine to conf/yarnconf of gearpump. $HADOOP_HOME points to the Hadoop installation directory.
```

Copying HADOOP config files to conf/yarnconf does not change the output. Since I'm launching on a YARN node this step shouldn't be necessary. HADOOP information is included below

```
$ yarn version
Hadoop 2.6.0-cdh5.4.2
Subversion http://github.com/cloudera/hadoop -r 15b703c8725733b7b2813d2325659eb7d57e7a3f
Compiled by jenkins on 2015-05-20T00:03Z
Compiled with protoc 2.5.0
From source with checksum de74f1adb3744f8ee85d9a5b98f90d
This command was run using /opt/cloudera/parcels/CDH-5.4.2-1.cdh5.4.2.p0.2/jars/hadoop-common-2.6.0-cdh5.4.2.jar
```

**Comment by kkasravi** [comment]
Note - when the appmaster launches the uiservice it adds a supervisor argument which is being set to localhost. This is incorrect. See io.gearpump.experiments.yarn.appmaster.UIService
```
  private val supervisor = ActorUtil.getFullPath(context.system, context.parent.path)
...
  Services.main(Array("-supervisor", supervisor, "-master", s"$masterHost:$masterPort"
```

**Comment by clockfly** [comment]
Thanks, confirmed it is a bug.

**Comment by clockfly** [comment]
fixed in PR #1818 
The number of initial workers can be configured in gear.conf
gearpump.yarn.worker.containers
